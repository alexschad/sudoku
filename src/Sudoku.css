.Sudoku {
  text-align: center;
  height: 100vh;
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.Sudoku-header {
  background-color: #282c34;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
  position: absolute;
  top: 0;
  width: 100%;
}

.container {
  display: flex;
  justify-content: center;
  height: 100%;
  flex-direction: column;
  align-items: center;
  background-color: paleturquoise;
}

.solving {
  opacity: 0.5;
}

.board {
  display: grid;
  grid-template:
    1fr 1fr 1fr 1fr 1fr 1fr 1fr 1fr 1fr / 
    1fr 1fr 1fr 1fr 1fr 1fr 1fr 1fr 1fr;
  height: 50vw;
  max-height: 800px;
  width: 50vw;
  max-width: 800px;
  margin: 20px;
}

.field {
  position: relative;
  background: rgb(237 255 91);
  border-left: 1px solid black;
  border-bottom: 1px solid black;
  display: flex;
  align-items: center;
  justify-content: center;
  color: blue;
  font-size: 2em;
}

.field.highlight {
  background: rgb(255, 225, 58);
}

.field.errHighlight {
  background: rgb(255, 175, 150);
}

.field.errNumber {
  color: red;
}

.field.fixed {
  color: black;
}

.top {
  border-top: 4px solid black;
  border-left: 1px solid black;
  border-bottom: 1px solid black;
}

.topleft {
  border-top: 4px solid black;
  border-left: 4px solid black;
  border-bottom: 1px solid black;
}

.topright {
  border-top: 4px solid black;
  border-right: 4px solid black;
  border-left: 1px solid black;
  border-bottom: 1px solid black;
}

.bottom {
  border-bottom: 4px solid black;
  border-left: 1px solid black;
}

.bottomleft {
  border-left: 4px solid black;
  border-bottom: 4px solid black;
}

.bottomright {
  border-bottom: 4px solid black;
  border-right: 4px solid black;
  border-left: 1px solid black;
}

.left {
  border-left: 4px solid black;
  border-bottom: 1px solid black;
}
.right {
  border-right: 4px solid black;
  border-left: 1px solid black;
  border-bottom: 1px solid black;
}

.field.over {
  background: rgb(248, 201, 72);
}

#numberSelectorContainer {
  position: absolute;
}

.field > .hint {
  font-size: calc(5px + 1vmin);
  color: red;
  position: absolute;
  top: 1px;
  right: 2px;
}

.numberSelector {
  display: grid;
  width: 15vw;
  max-width: 200px;
  height: 15vw;
  max-height: 200px;
  background-color: white;
  border: 1px solid black;
  font-size: calc(10px + 2vmin);
  grid-template:
    1fr 1fr 1fr  / 
    1fr 1fr 1fr;
}

.numberSelector > div {
  border-left: 1px solid black;
  border-bottom: 1px solid black;
  display: flex;
  align-items: center;
  justify-content: center;
}

.numberSelector > div:hover {
  background-color: aquamarine;
}

.selectorClose {
  position: absolute;
  background-color: white;
  border: 1px solid black;
  border-radius: 50%;
  width: 0.8em;
  height: 0.8em;
  font-size: .5em;
  padding: 5px;
  top: -0.5em;
  left: -0.5em;
  display: flex;
  align-items: center;
  justify-content: center;
}

.selectorClose:hover {
  background-color: aquamarine;
}

.gone {
  display: none;
}

.code {
  width: 70vw;
  overflow-wrap: anywhere;
}

.success {
  position: absolute;
  color: #55e445;
  font-size: 3em;
  font-weight: bold;
  height: 2em;
  z-index: 1;
  background-blend-mode: exclusion;
  width: 100vw;
  height: 100vh;
  display: flex;
  align-items: center;
  justify-content: center;
  background: rgb(109 101 0 / 79%);
}

button {
    box-shadow: inset 0px 1px 0px 0px #fff6af;
    background: linear-gradient(to bottom, #ffec64 5%, #ffab23 100%);
    background-color: #ffec64;
    border-radius: 6px;
    border: 1px solid #ffaa22;
    display: inline-block;
    cursor: pointer;
    color: #333333;
    font-family: Arial;
    font-size: 15px;
    font-weight: bold;
    padding: 6px 24px;
    text-decoration: none;
    text-shadow: 0px 1px 0px #ffee66;
}

button:hover {
	background:linear-gradient(to bottom, #ffab23 5%, #ffec64 100%);
	background-color:#ffab23;
}

.menu {
  margin: 20px;
  display: flex;
  width: 60%;
}

.menu > * {
  flex-basis: 100%;
  margin: 0px 5px;
  outline: none;
}