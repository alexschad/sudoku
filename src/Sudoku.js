import React, { useState, useEffect, useRef } from 'react';
import './Sudoku.css';
import Field from './Field';
import NumberSelector from './NumberSelector';
import { rows, cols, squares } from './data';
import easy from './easy';
import medium from './medium';
import hard from './hard';
import { checkValid, formatTime } from './util';
import {solve } from './solve';
import NewGameDialog from './NewGameDialog';

import { styled } from "@material-ui/core/styles";
import { spacing } from "@material-ui/system";
import MuiButton from "@material-ui/core/Button";

/**
 * Applies the spacing system to the material UI Button
 */
const Button = styled(MuiButton)(spacing);


export const ACTIONS = {
  LOAD_RANDOM: 'load_random',
  LOAD: 'load',
  CLEAR: 'clear',
  RESET: 'reset',
  SOLVE: 'solve',
  SOLVE_ONE: 'solve_one',
  SET_FIELD: 'set_field',
  PAUSE_GAME: 'pause_game',
  SET_TIMER: 'set_timer',
};

const SUDOKUS = {
  easy,
  medium,
  hard,
}

const sudokuReducer = (state, action) => {
  const { type, fields, sudokuIndex } = state;
  switch (action.type) {
    // randomly loads one of the stored sudokus
    case ACTIONS.LOAD_RANDOM:
      let newType;
      let newIndex;
      let newFields;
      if (action.payload.type === 'random') {
        newType = ['easy', 'medium', 'hard'][Math.floor(Math.random() * 3)];
        newIndex = Math.floor(Math.random() * SUDOKUS[newType].length);
        newFields = SUDOKUS[newType][newIndex];
      } else if (action.payload.type === 'custom') {
        newType = 'custom';
        newIndex = null;
        newFields = Array(81).fill(0);
      } else {
        newType = action.payload.type;
        newIndex = Math.floor(Math.random() * SUDOKUS[newType].length);
        newFields = SUDOKUS[newType][newIndex];
      }
      return {...state, type: newType, fields: newFields, sudokuIndex: newIndex, timer: 0, gameState: 'running'};
    // loads one of the stored sudokus
    case ACTIONS.LOAD:
      return {...state, fields: SUDOKUS[type][action.payload.sudokuIndex], sudokuIndex: action.payload.sudokuIndex, timer: 0};
    // clears all user selected fields. resets to the current sudoku
    case ACTIONS.RESET:
      if (!type || type === 'custom') {
        return {...state, type: 'custom', fields: Array(81).fill(0), sudokuIndex: null, timer: 0};
      }
      return {...state, fields: SUDOKUS[type][sudokuIndex], sudokuIndex: sudokuIndex, timer: 0};
    // clears all fields
    case ACTIONS.CLEAR:
      return {...state, fields: Array(81).fill(0), sudokuIndex: null, timer: 0};
    // solve the sudoku
    case ACTIONS.SOLVE:
      const solved = solve(fields.join());
      return {...state, fields: solved};
    // solve one field of the sudoku
    case ACTIONS.SOLVE_ONE:
      const ones = fields.map((field,index) => {
        const row = rows.find(r => r.includes(index));
        const col = cols.find(c => c.includes(index));
        const square = squares.find(s => s.includes(index));
        const usedIndices = (row && col && square) ? row.concat(col).concat(square) : [];
        let usedVals = usedIndices.map(ui => fields[ui]).filter(i => i !== 0);
        usedVals = new Set(usedVals);  
        usedVals = [...usedVals];
        return ([index, usedVals]);
      }).filter(
        (o) => {
          return o[1].length === 8 && fields[o[0]] === 0;
        }
      );
      if (ones.length > 0) {
        // try a smart solution
        const sol = ones[Math.floor(Math.random() * ones.length)];
        sol[1] = [1,2,3,4,5,6,7,8,9].filter(x => !sol[1].includes(x))[0];
        const newFields = [...fields];
        newFields[sol[0]] = sol[1];
        return {...state, fields: newFields};
      } else {
        const solved = solve(fields.join());
        const nullIndices = fields.map((f,i) => i).filter(i => fields[i] === 0);
        const solIndex = nullIndices[Math.floor(Math.random() * nullIndices.length)];
        const sol = [solIndex, solved[solIndex]];
        const newFields = [...fields];
        newFields[sol[0]] = sol[1];
        return {...state, fields: newFields};
      }
    // set one field with a number
    case ACTIONS.SET_FIELD:
      const currentFields = [...fields];
      currentFields[action.payload.clicked] = action.payload.value;
      return {...state, fields: currentFields};
    // pause game
    case ACTIONS.PAUSE_GAME:
      return {...state, gameState: 'paused'};
    // resume game
    case ACTIONS.RESUME_GAME:
      return {...state, gameState: 'running'};
    case ACTIONS.SET_TIMER:
      return {...state, timer: action.payload.timer};
    default:
      return state;
  }
};

//main Module that shows the Fields, the Menu and the Number selector
function Sudoku() {
  const timeOutRef = useRef(null);
  const initialSudoku = () => {
    const sudokuTypeJSON = window.localStorage.getItem("type");
    const type = sudokuTypeJSON ? JSON.parse(sudokuTypeJSON) : null;
    const sudokuIndexJSON = window.localStorage.getItem("index");
    const sudokuIndex = sudokuIndexJSON ? JSON.parse(sudokuIndexJSON) : null;
    const sudokuFieldsJSON = window.localStorage.getItem("fields");
    const fields = sudokuFieldsJSON ? JSON.parse(sudokuFieldsJSON) : Array(81).fill(0);
    const sudokuStateJSON = window.localStorage.getItem("gameState");
    const gameState = sudokuStateJSON ? JSON.parse(sudokuStateJSON) : 'paused';
    const sudokuTimerJSON = window.localStorage.getItem("timer");
    const timer = sudokuTimerJSON ? JSON.parse(sudokuTimerJSON) : 0;
    return {type, fields, sudokuIndex, gameState, timer};
  };

  const [{type, fields, sudokuIndex, gameState, timer}, dispatch] = React.useReducer(
    sudokuReducer,
    undefined,
    initialSudoku
  );
  const [highlight, setHightlight] = useState([]);
  const [errHighlight, setErrHightlight] = useState([]);
  const [errNumber, setErrNumber] = useState([]);
  const [over, setOver] = useState();
  const [clicked, setClicked] = useState(null);
  const [selPos, setSelPos] = useState([0,0]);
  const [isSolving, setIsSolving] = useState(false);
  const [isSolved, setIsSolved] = useState(false);
  const [showHints, setShowHints] = useState(false);
  const [open, setOpen] = React.useState(type === null);

  const isFirstRenderRef = useRef(true);

  useEffect(() => window.localStorage.setItem('type', JSON.stringify(type)), [type]);
  useEffect(() => window.localStorage.setItem('fields', JSON.stringify(fields)), [fields]);
  useEffect(() => {
    clearTimeout(timeOutRef.current);
    window.localStorage.setItem('gameState', JSON.stringify(gameState));
  }, [gameState]);
  useEffect(() => {
    if (timer === 0) {
      clearTimeout(timeOutRef.current);
    }
    if (gameState === 'running') {
      timeOutRef.current = setTimeout(
        () => {
          dispatch({ type: ACTIONS.SET_TIMER, payload: { timer: timer + 1 }});
          window.localStorage.setItem('timer', JSON.stringify(timer))
        }, 1000);
    }
  }, [timer, gameState]);



  useEffect(() => {
    if (isFirstRenderRef.current) return;
    if (sudokuIndex !== null) {
      dispatch({ type: ACTIONS.LOAD, payload: { sudokuIndex }})
    } else {
      dispatch({ type: ACTIONS.CLEAR })
    }
    window.localStorage.setItem('index', JSON.stringify(sudokuIndex));
  }, [sudokuIndex, isFirstRenderRef]);

  useEffect(() => {
    isFirstRenderRef.current = false;
  }, []);

  // start to solve the whole sudoku
  const startSolve = async () => {
    setIsSolving(true);
  }

  useEffect(() => {
    if (isSolving === false) return;
    dispatch({ type: ACTIONS.SOLVE })
    setIsSolving(false);
  }, [isSolving, fields]);

  useEffect(() => {
    const solved = fields.filter(i => i === 0).length === 0;
    setIsSolved(solved);
  }, [fields]);

  // set the row, column and block highlights on mouseover
  const mouseEnter = (index) => {
    const newHighlight = [
      ...rows.find(r => r.includes(index)),
      ...cols.find(r => r.includes(index)),
      ...squares.find(r => r.includes(index)),
    ];
    setHightlight(newHighlight);
    setOver(index);
  };
  
  // resets the highlights
  const mouseLeave = () => {
    setHightlight([]);
    setOver(null);
  };


  // sets a number in a field called by the number selector
  const selectNumber = (clicked, value) => {
    dispatch({ type: ACTIONS.SET_FIELD, payload: { clicked, value }});
    setClicked(null);
  }

  // shows and places the number selector
  const showSelector = (e, index) => {
    setSelPos([e.clientX, e.clientY]);
    setClicked(index);
  }

  useEffect(() => {
    const [errn, errh] = checkValid(fields);
    setErrNumber(errn);
    setErrHightlight(errh);
  }, [fields]);

  // if (gameState === 'paused') {
  //   return (<div className="Menu">
  //     <h3>Game Paused</h3>
  //     <button onClick={() => dispatch({ type: ACTIONS.LOAD_RANDOM })}>Random</button>
  //   </div>);
  // }

  const handleClickOpen = () => {
    setOpen(true);
  };

  const handleClose = (value) => {
    dispatch({ type: ACTIONS.LOAD_RANDOM, payload: { type: value } })
    setOpen(false);
  };

  return (
    <div className="Sudoku">
      <div className={isSolving ? 'container solving' : 'container'}>
      <header className="Sudoku-header">Sudoku {type ? `(${type})` : null}</header>
      {isSolved ? <div className="success" onClick={() => setIsSolved(false)}>Congratulations you solved the sudoku!</div> : null}
      {gameState === 'paused' ? 
        <div className="pausedBoard" onClick={() => dispatch({ type: ACTIONS.RESUME_GAME })}>
          {Array(81).fill(0).map((f,i) => 
            <Field
              key={i}
              index={i}
              fields={Array(81).fill(0)}
              sudoku={null}
              showSelector={() => {}}
              showHints={false}
              mouseEnter={() => {}}
              mouseLeave={() => {}}
            />)}
        </div> : 
        <div className="board">
          {fields.map((f,i) => 
            <Field
              key={i}
              index={i}
              fields={fields}
              sudoku={type !== null && sudokuIndex!== null ? SUDOKUS[type][sudokuIndex] : null}
              showSelector={showSelector}
              showHints={showHints}
              mouseEnter={mouseEnter}
              mouseLeave={mouseLeave}
              highlight={highlight.includes(i)}
              errHighlight={errHighlight.includes(i)}
              errNumber={errNumber.includes(i)}
              over={over === i}
            />
          )}
        </div>
        }
        <NewGameDialog  open={open} onClose={handleClose} />
        <div className="timerContainer">
            <p className="timer">{formatTime(timer)}</p>
            {gameState === 'running' ? <div className="pause" onClick={() => dispatch({ type: ACTIONS.PAUSE_GAME })} /> : <div className="resume" onClick={() => dispatch({ type: ACTIONS.RESUME_GAME })} />}
          </div>
        <div className="menu">

          <Button mx={1} onClick={handleClickOpen}>New Game</Button>
          <Button mx={1} onClick={startSolve}>Solve Sudoku</Button>
          <Button mx={1} onClick={() => dispatch({ type: ACTIONS.SOLVE_ONE })}>Solve One</Button>
          <Button mx={1} onClick={() => dispatch({ type: ACTIONS.RESET })}>Reset</Button>
          <Button mx={1} onClick={(e) => setShowHints(!showHints)}>{showHints ? 'Hide' : 'Show'} Hints</Button>
        </div>
        <NumberSelector
          selectNumber={selectNumber}
          clicked={clicked}
          setClicked={setClicked}
          selPos={selPos}
          fields={fields}
        />
      </div>
    </div>
  );
}

export default Sudoku;
