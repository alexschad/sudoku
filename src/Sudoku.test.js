import React from 'react';
import renderer from 'react-test-renderer';
import Sudoku from './Sudoku';

test('it renders', () => {
  const component = renderer.create(
    <Sudoku />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

});