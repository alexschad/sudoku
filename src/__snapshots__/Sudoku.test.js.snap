// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<div
  className="Sudoku"
>
  <div
    className="container"
  >
    <header
      className="Sudoku-header"
    >
      Sudoku Helper
    </header>
    <div
      className="board"
    >
      <div
        className="field topleft"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field top"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field top"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field topleft"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field top"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field top"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field topleft"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field top"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field topright"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field left"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field left"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field left"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field right"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field left"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field left"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field left"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field right"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field topleft"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field top"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field top"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field topleft"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field top"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field top"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field topleft"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field top"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field topright"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field left"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field left"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field left"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field right"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field left"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field left"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field left"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field right"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field topleft"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field top"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field top"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field topleft"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field top"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field top"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field topleft"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field top"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field topright"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field left"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field left"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field left"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field right"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field bottomleft"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field bottom"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field bottom"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field bottomleft"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field bottom"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field bottom"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field bottomleft"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field bottom"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
      <div
        className="field bottomright"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </div>
    </div>
    <div
      className="menu"
    >
      <button
        onClick={[Function]}
      >
        Load
      </button>
      <button
        onClick={[Function]}
      >
        Start Solving
      </button>
      <button
        onClick={[Function]}
      >
        Solve One
      </button>
      <button
        onClick={[Function]}
      >
        Clear
      </button>
      <button
        onClick={[Function]}
      >
        Reset
      </button>
      <button
        onClick={[Function]}
      >
        Show
         Hints
      </button>
    </div>
    <div
      id="numberSelectorContainer"
      style={
        Object {
          "display": "none",
          "left": "0px",
          "top": "0px",
        }
      }
    >
      <div
        className="numberSelector"
        onMouseLeave={[Function]}
      >
        <div
          className="selectorClose"
          onClick={[Function]}
        >
          X
        </div>
        <div
          onClick={[Function]}
        >
          1
        </div>
        <div
          onClick={[Function]}
        >
          2
        </div>
        <div
          onClick={[Function]}
        >
          3
        </div>
        <div
          onClick={[Function]}
        >
          4
        </div>
        <div
          onClick={[Function]}
        >
          5
        </div>
        <div
          onClick={[Function]}
        >
          6
        </div>
        <div
          onClick={[Function]}
        >
          7
        </div>
        <div
          onClick={[Function]}
        >
          8
        </div>
        <div
          onClick={[Function]}
        >
          9
        </div>
      </div>
    </div>
  </div>
</div>
`;
